doc/holochain_101/src/json_string.md

```diff
@@ -0,0 +1,823 @@
+# Serialization and JsonString
+
+## Why serialize anything? Why JSON?
+
+### Holochain zomes are written in WASM.
+
+WASM only supports working directly with integers and manually allocating
+memory. This means that sharing any data between holochain core and zome
+functions must be serialized. There is no way that WASM functions can
+understand the Rust type system natively. Serialized data can be allocated for
+WASM to read out and deserialize into Rust types/structs/enums.
+
+Any developers using the Rust HDK get the serialization/deserialization and
+type handling _almost_ "for free". The macros for defining entities and zomes
+automatically wrap the memory work and serialization round trips for anything
+that implements `Into<JsonString>` and `TryFrom<JsonString>` (see below).
+
+We use `serde` for our serialization round trips as it is by far the most
+popular and mature option for Rust. Many serialization formats other than JSON
+are supported by `serde` but JSON is a solid option. JSON allows us to easily
+bring the Rust type system across to WASM with decent performance.
+
+From the `serde_json` github repository README:
+
+> It is fast. You should expect in the ballpark of 500 to 1000 megabytes per
+> second deserialization and 600 to 900 megabytes per second serialization,
+> depending on the characteristics of your data. This is competitive with the
+> fastest C and C++ JSON libraries or even 30% faster for many use cases.
+> Benchmarks live in the serde-rs/json-benchmark repo.
+
+### Holochain aims to support all WASM languages not just Rust/JS
+
+The official Holochain HDKs are Rust and AssemblyScript. The Rust HDK will
+always be the most tightly integrated HDK with core simply because Holochain
+itself is Rust based.
+
+Generally though, we are hoping and expecting many different WASM zome
+languages build an ecosystem over time. Personally I'm hoping for a decent LISP
+to appear ;)
+
+To encourage as many languages as possible we want to keep the minimum
+requirements for interacting with holochain core as minimal as possible.
+
+Currently the two requirements for writing zomes in `<your favourite language>`:
+
+- Must compile to WASM
+- Must be able to serialize UTF-8 data and allocate to memory read by core
+
+We can't do much about the first requirement but here are some lists to watch:
+
+- https://github.com/appcypher/awesome-wasm-langs
+- https://github.com/mbasso/awesome-wasm
+
+The second requirement means that we must be very mindful of choosing a
+serialization format that can round trip through as many languages as possible.
+
+In the end, this is the main reason we chose JSON for communication with core.
+
+Note that when we started on an AssemblyScript (ostentisbly JavaScript) HDK
```

Comment:
@thedavidmeister there still isn't ... we never got what we worked on merged upstream!